In this project I tried to use and develop my pandas skills.I did some basic cleaning and analysis tasks.    

# Exploring eBay Car Sales Data
dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
The data dictionary provided with data is as follows:

dateCrawled - When this ad was first crawled. All field-values are taken from this date.
name - Name of the car.
seller - Whether the seller is private or a dealer.
offerType - The type of listing
price - The price on the ad to sell the car.
abtest - Whether the listing is included in an A/B test.
vehicleType - The vehicle Type.
yearOfRegistration - The year in which the car was first registered.
gearbox - The transmission type.
powerPS - The power of the car in PS.
model - The car model name.
odometer - How many kilometers the car has driven.
monthOfRegistration - The month in which the car was first registered.
fuelType - What type of fuel the car uses.
brand - The brand of the car.
notRepairedDamage - If the car has a damage which is not yet repaired.
dateCreated - The date on which the eBay listing was created.
nrOfPictures - The number of pictures in the ad.
postalCode - The postal code for the location of the vehicle.
lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [2]:
import pandas as pd
import numpy as np


In [3]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [5]:
autos 

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset contains 20 columns, most of which are strings.
Some columns have null values, but none have more than ~20% null values.
The column names use camelcase instead of Python's preferred snakecase
Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [5]:
# Installing inflection library to convert camelCase to snake_case
import sys
!{sys.executable} -m pip install inflection
import inflection

You should consider upgrading via the '/dataquest/system/env/python3/bin/python3 -m pip install --upgrade pip' command.


In [6]:
inflection.underscore("exampleCamelCase")

'example_camel_case'

In [7]:
new_column_names = []

for names in autos.columns:
    if names == "yearOfRegistration":
        names = "registration_year"
    elif names == "monthOfRegistration":
        names = "registration_month"
    elif names == "notRepairedDamage ":
        names = "unrepaired_damage"
    elif names == "dateCreated":
         names = "ad_created"
            
    names = inflection.underscore(names)
            
    new_column_names.append(names)

autos.columns = new_column_names
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand not_repaired_damage           ad_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

I changed the column names to be more descriptive and easy_to_read. 

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand  \
count       50000        50000.000000     45518       50000   
unique         13                 NaN         7          40   
top     150,000km                 NaN    benzin  volkswagen   
freq        32424                 NaN     30107       10687   
mean          NaN            5.723360       NaN         NaN   
std           NaN            3.711984       NaN         NaN   
min           NaN            0.000000       NaN         NaN   
25%           NaN            3.000000       NaN         NaN   
50%           NaN            6.000000       NaN         NaN   
75%           NaN            9.000000       NaN         NaN   
max           NaN           12.000000       NaN         NaN   

       not_repaired_damage           ad_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Any columns that have mostly one value that are candidates to be dropped
Any columns that need more investigation.
Any examples of numeric data stored as text that needs to be cleaned.

As can be seen from previous table there are no numerical results for price and odometer columns. for example $5000 and 150000km values can be represented in numerical format. 

So I will remove non-numeric characters and convert them into numeric dtype, and rename odemeter column as odometer_km to show it'n unit as km.  

In [9]:
autos["price"] = autos["price"].str.replace('$','').str.replace(',','').astype(int)
print(autos["price"][:5])

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64


In [10]:
autos["odometer"] = autos["odometer"].str.replace(',','').str.replace('km','').astype(int)
print(autos["odometer"][:5])

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64


In [11]:
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [12]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'not_repaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [13]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                 50000        50000   50000      50000  5.000000e+04   
unique                48213        38754       2          2           NaN   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           78   49999      49999           NaN   
mean                    NaN          NaN     NaN        NaN  9.840044e+03   
std                     NaN          NaN     NaN        NaN  4.811044e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.100000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       abtest vehicle_type  registration_year  gearbox      power_ps  model  \
count   50000        44905       50000.000000    47320  50000.000000  47242   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    25756        12859                NaN    36993           NaN   4024   
mean      NaN          NaN        2005.073280      NaN    116.355920    NaN   
std       NaN          NaN         105.712813      NaN    209.216627    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%       NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    50000.000000        50000.000000     45518       50000   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     30107       10687   
mean    125732.700000            5.723360       NaN         NaN   
std      40042.211706            3.711984       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       not_repaired_damage           ad_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

We checked for odometer and price columns and converted them into int type. Now we'll explore further and look for outliers if there is any. 

In [16]:
print(autos["odometer_km"].unique().shape)
autos["odometer_km"].describe()

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
autos["odometer_km"].value_counts().sort_index(ascending=False).head()
# I am checking for number of values for corresponding km's, sorting them in descending order
# and displaying first 5 rows

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64

Since these cars used we can assume that odometer values looks ok. I don't think any interesting value exist.

I am searching for price column as well if any unusual situtation exists

In [20]:
print(autos["price"].unique().shape)
autos["price"].describe()

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [22]:
autos["price"].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price, Length: 2357, dtype: int64

In [23]:
print(autos["price"].median())
# Median looks ok but mean might be skewed 

2950.0


I can see some unrealstic numbers in price column. For example there 1421 cars with zero cost.let's search a little bit

In [33]:
autos.loc[autos["price"] == 11111111,"name"]

2897     Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000
24384                             Schlachte_Golf_3_gt_tdi
Name: name, dtype: object

When I check for escort MK 1 on the internet I saw price tag as £47,995 so we can be sure about that values seem like 11111111 or equal number of digits are weird and we can eliminate them.  
https://www.carandclassic.com/list/21/escort+mk1+rs2000/

I will keep price tags between 250 and 50,000

In [37]:
autos_test_3 = autos[autos["price"].between(250,50000)]
print(autos_test_3.shape)
print(autos_test_3["price"].value_counts().sort_index(ascending=False))

(47102, 20)
50000      2
49999      4
49900      5
49500      4
49000      3
        ... 
269        1
260        5
255        1
251        1
250      291
Name: price, Length: 2124, dtype: int64


seems more reasonable so I can assign my test df to original autos df 

In [38]:
autos = autos[autos["price"].between(250,50000)]
print(autos.shape)

(47102, 20)


I will check for columns that provides information about time.


In [39]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 47102 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   date_crawled         47102 non-null  object
 1   name                 47102 non-null  object
 2   seller               47102 non-null  object
 3   offer_type           47102 non-null  object
 4   price                47102 non-null  int64 
 5   abtest               47102 non-null  object
 6   vehicle_type         42947 non-null  object
 7   registration_year    47102 non-null  int64 
 8   gearbox              45054 non-null  object
 9   power_ps             47102 non-null  int64 
 10  model                44823 non-null  object
 11  odometer_km          47102 non-null  int64 
 12  registration_month   47102 non-null  int64 
 13  fuel_type            43470 non-null  object
 14  brand                47102 non-null  object
 15  not_repaired_damage  38657 non-null  object
 16  ad_c

date_crawled , last_seen and ad_created columns are string. registration_year and registration_month calues are integer. 

In [40]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

I dont need hours and first 10 characters represent the day. I can take it

In [58]:
autos["date_crawled"].str[:10].value_counts(dropna = False).sort_values(ascending=False)
"""autos["date_crawled"].str[:10].value_counts(dropna = False).sort_index(ascending=False) sort by index or values ;)  """ 

'autos["date_crawled"].str[:10].value_counts(dropna = False).sort_index(ascending=False) sort by index or values ;)  '

In [56]:
autos["ad_created"].str[:10].value_counts(dropna = False).sort_values(ascending=False)

2016-04-03    1838
2016-03-20    1797
2016-03-21    1771
2016-04-04    1740
2016-03-12    1737
              ... 
2016-02-22       1
2016-02-11       1
2016-02-01       1
2015-08-10       1
2015-11-10       1
Name: ad_created, Length: 76, dtype: int64

In [57]:
autos["last_seen"].str[:10].value_counts(dropna = False).sort_values(ascending=False)


2016-04-06    10482
2016-04-07     6238
2016-04-05     5902
2016-03-17     1325
2016-04-03     1186
2016-04-02     1171
2016-04-04     1160
2016-03-30     1159
2016-03-12     1128
2016-03-31     1123
2016-04-01     1082
2016-03-29     1037
2016-03-22     1012
2016-03-28      982
2016-03-20      970
2016-03-21      967
2016-03-24      924
2016-03-25      893
2016-03-23      870
2016-03-26      784
2016-03-16      768
2016-03-15      744
2016-03-19      735
2016-03-27      732
2016-03-14      589
2016-03-11      583
2016-03-10      496
2016-03-09      452
2016-03-13      422
2016-03-18      341
2016-03-08      339
2016-03-07      252
2016-03-06      204
2016-03-05       50
Name: last_seen, dtype: int64

In [53]:
autos["registration_year"].describe()

count    47102.000000
mean      2004.842661
std         88.923813
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [61]:
autos["registration_year"].value_counts().sort_index(ascending=False)

9999    3
9000    1
8888    1
6200    1
5911    1
       ..
1910    2
1800    2
1111    1
1001    1
1000    1
Name: registration_year, Length: 94, dtype: int64

Obviously we have some problems here. Year of registiration should be clean. we have values from year 9999 and 1001 which are probably not true. so I will restrict with the years 1900 to 2016 I think there could be some really old cars this is why I set boundry as 1900 and latest models are probably 2016 and earlier. 


In [62]:
autos = autos[autos["registration_year"].between(1900,2016)]
print(autos.shape)

(45254, 20)


Now I want to explore brand column. I can see distrubution of cars by brand by following lines. Then I will calculate avarage prices for each brand.  

In [67]:
print(autos["brand"].value_counts())

volkswagen        9600
bmw               5068
opel              4767
mercedes_benz     4437
audi              3983
ford              3084
renault           2093
peugeot           1359
fiat              1134
seat               827
skoda              755
nissan             691
mazda              689
smart              658
citroen            642
toyota             592
hyundai            461
volvo              421
sonstige_autos     421
mini               406
mitsubishi         371
honda              361
kia                327
alfa_romeo         302
suzuki             265
chevrolet          261
porsche            195
chrysler           162
dacia              123
daihatsu           108
jeep               105
subaru              94
land_rover          92
saab                75
jaguar              68
daewoo              63
rover               61
trabant             58
lancia              48
lada                27
Name: brand, dtype: int64


' first of all I need to keep unique values of brands so that I can aggregate'

In [68]:
""" first of all I need to keep unique values of brands so that I can aggregate"""
unique_brands = autos["brand"].unique()
unique_brands
#

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'hyundai', 'chevrolet', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [78]:
""" now I can create an empty dictionary to store avarage prices"""
avg_brand_price = {}

for each in unique_brands:
    average = autos.loc[autos["brand"]== each,"price"].mean()
    avg_brand_price[each] = average
    
print(avg_brand_price)

{'peugeot': 3167.9941133186167, 'bmw': 8181.520323599053, 'volkswagen': 5532.8090625, 'smart': 3596.40273556231, 'ford': 3815.6854734111544, 'chrysler': 3506.864197530864, 'seat': 4531.4691656590085, 'renault': 2550.726230291448, 'mercedes_benz': 8257.55262564796, 'audi': 9105.847602309817, 'sonstige_autos': 9923.65083135392, 'opel': 3126.904552129222, 'mazda': 4227.840348330915, 'porsche': 23866.584615384614, 'mini': 10691.06157635468, 'toyota': 5175.555743243243, 'dacia': 5915.528455284553, 'nissan': 4889.322720694646, 'jeep': 11230.028571428571, 'saab': 3292.6266666666666, 'volvo': 5016.275534441806, 'mitsubishi': 3509.730458221024, 'jaguar': 10840.764705882353, 'fiat': 2962.0123456790125, 'skoda': 6459.275496688741, 'subaru': 4197.95744680851, 'kia': 6036.388379204893, 'citroen': 3847.1105919003116, 'hyundai': 5445.095444685467, 'chevrolet': 6594.827586206897, 'honda': 4163.08864265928, 'daewoo': 1147.936507936508, 'suzuki': 4276.698113207547, 'trabant': 1990.551724137931, 'land_ro

In [82]:
sorted_avg_brand_price = sorted(avg_brand_price.items(), key=lambda x:x[1], reverse=True)
print(sorted_avg_brand_price)

[('porsche', 23866.584615384614), ('land_rover', 15481.445652173914), ('jeep', 11230.028571428571), ('jaguar', 10840.764705882353), ('mini', 10691.06157635468), ('sonstige_autos', 9923.65083135392), ('audi', 9105.847602309817), ('mercedes_benz', 8257.55262564796), ('bmw', 8181.520323599053), ('chevrolet', 6594.827586206897), ('skoda', 6459.275496688741), ('kia', 6036.388379204893), ('dacia', 5915.528455284553), ('volkswagen', 5532.8090625), ('hyundai', 5445.095444685467), ('toyota', 5175.555743243243), ('volvo', 5016.275534441806), ('nissan', 4889.322720694646), ('seat', 4531.4691656590085), ('suzuki', 4276.698113207547), ('mazda', 4227.840348330915), ('subaru', 4197.95744680851), ('honda', 4163.08864265928), ('alfa_romeo', 4010.0496688741723), ('citroen', 3847.1105919003116), ('ford', 3815.6854734111544), ('smart', 3596.40273556231), ('lancia', 3512.5833333333335), ('mitsubishi', 3509.730458221024), ('chrysler', 3506.864197530864), ('saab', 3292.6266666666666), ('peugeot', 3167.994113

In [88]:
avg_brand_series = pd.Series(avg_brand_price)
avg_brand_series.sort_values(ascending=False)

porsche           23866.584615
land_rover        15481.445652
jeep              11230.028571
jaguar            10840.764706
mini              10691.061576
sonstige_autos     9923.650831
audi               9105.847602
mercedes_benz      8257.552626
bmw                8181.520324
chevrolet          6594.827586
skoda              6459.275497
kia                6036.388379
dacia              5915.528455
volkswagen         5532.809063
hyundai            5445.095445
toyota             5175.555743
volvo              5016.275534
nissan             4889.322721
seat               4531.469166
suzuki             4276.698113
mazda              4227.840348
subaru             4197.957447
honda              4163.088643
alfa_romeo         4010.049669
citroen            3847.110592
ford               3815.685473
smart              3596.402736
lancia             3512.583333
mitsubishi         3509.730458
chrysler           3506.864198
saab               3292.626667
peugeot            3167.994113
opel    

Now I would like to store top 6 preffered brands. 

In [98]:
top_brands = autos["brand"].value_counts()[:7]
top_brands.index[0]

'volkswagen'

In [100]:
top_7_brands = top_brands.index[:7]
print(top_7_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford',
       'renault'],
      dtype='object')


In [101]:
avg_brand_price_7 = {}
for brand in top_7_brands:
    avg_price = autos.loc[autos["brand"] == brand,"price"].mean()
    avg_brand_price_7[brand] = avg_price

print(avg_brand_price_7)

{'volkswagen': 5532.8090625, 'bmw': 8181.520323599053, 'opel': 3126.904552129222, 'mercedes_benz': 8257.55262564796, 'audi': 9105.847602309817, 'ford': 3815.6854734111544, 'renault': 2550.726230291448}


In [102]:
sorted_top_7 = sorted(avg_brand_price_7.items(), key=lambda x:x[1], reverse = True)
print(sorted_top_7)

[('audi', 9105.847602309817), ('mercedes_benz', 8257.55262564796), ('bmw', 8181.520323599053), ('volkswagen', 5532.8090625), ('ford', 3815.6854734111544), ('opel', 3126.904552129222), ('renault', 2550.726230291448)]


Audi is the most expensiv eone among most preffered brands. We found that  porsche has the highest avarage with 23866.584615 but it is not one of top brands. 